perfkitbenchmarker/windows_virtual_machine.py

```diff
@@ -0,0 +1,306 @@
+      package.Install(self)
+      self._installed_packages.add(package_name)
+
+  def Uninstall(self, package_name):
+    """Uninstalls a Perfkit package on the VM."""
+    package = windows_packages.PACKAGES[package_name]
+    if hasattr(package, 'Uninstall'):
+      package.Uninstall()
+
+  def PackageCleanup(self):
+    """Cleans up all installed packages.
+
+    Deletes the temp directory and uninstalls all PerfKit packages.
+    """
+    for package_name in self._installed_packages:
+      self.Uninstall(package_name)
+    self.RemoteCommand('rm -recurse -force %s' % self.temp_dir)
+    self.DisableGuestFirewall()
+
+  def _GetNumCpus(self):
+    """Returns the number of logical CPUs on the VM.
+
+    This method does not cache results (unlike "num_cpus").
+    """
+    stdout, _ = self.RemoteCommand(
+        'Get-WmiObject -class Win32_processor | '
+        'select -exp NumberOfLogicalProcessors')
+    return int(stdout)
+
+  def _GetTotalMemoryKb(self):
+    """Returns the amount of physical memory on the VM in Kilobytes.
+
+    This method does not cache results (unlike "total_memory_kb").
+    """
+    stdout, _ = self.RemoteCommand(
+        'Get-WmiObject -class Win32_PhysicalMemory | '
+        'select -exp Capacity')
+    return int(stdout) / 1024
+
+  def _TestReachable(self, ip):
+    """Returns True if the VM can reach the ip address and False otherwise."""
+    try:
+      self.RemoteCommand('ping -n 1 %s' % ip)
+    except errors.VirtualMachine.RemoteCommandError:
+      return False
+    return True
+
+  def DownloadFile(self, url, dest):
+    """Downloads the content at the url to the specified destination."""
+
+    command = 'Invoke-WebRequest {url} -OutFile {dest}'.format(
+        url=url, dest=dest)
+    self.RemoteCommand(command)
+
+  def UnzipFile(self, zip_file, dest):
+    """Unzips the file with the given path."""
+    command = ('Add-Type -A System.IO.Compression.FileSystem; '
+               '[IO.Compression.ZipFile]::ExtractToDirectory(\'{zip_file}\', '
+               '\'{dest}\')').format(zip_file=zip_file, dest=dest)
+    self.RemoteCommand(command)
+
+  def DisableGuestFirewall(self):
+    """Disables the guest firewall."""
+    command = 'netsh advfirewall set allprofiles state off'
+    self.RemoteCommand(command)
+
+  def EnableGuestFirewall(self):
+    """Enables the guest firewall."""
+    command = 'netsh advfirewall set allprofiles state on'
+    self.RemoteCommand(command)
+
+  def _RunDiskpartScript(self, script):
+    """Runs the supplied Diskpart script on the VM."""
+    with vm_util.NamedTemporaryFile(prefix='diskpart') as tf:
+      tf.write(script)
+      tf.close()
+      self.RemoteCopy(tf.name, self.temp_dir)
+      script_path = os.path.join(self.temp_dir, os.path.basename(tf.name))
+      self.RemoteCommand('diskpart /s {script_path}'.format(
+          script_path=script_path))
+
+  def _CreateScratchDiskFromDisks(self, disk_spec, disks):
+    """Helper method to prepare data disks.
```

Comment:
There is a version of this without the need for disk numbers here: http://www.karafilis.net/windows-2012-storage-virtualization/
He queries for disks that can pool, which excludes the OS disk.